regexp: max \U and \%U value is limited by INT_MAX

chrisbra opened the pull request:
Problem:  regexp: max \U and \%U value is limited by INT_MAX but gives a
          confusing error message (related: v8.1.0985).
Solution: give a better error message when the value reaches INT_MAX

When searching Vim allows to get up to 8 hex characters using the /\V and /\%V regex atoms.  However, when using "/\UFFFFFFFF" the code point is already above what an integer variable can hold, which is 2,147,483,647.

So since patch v8.1.0985, Vim already limited the max codepoint to INT_MAX (otherwise it caused a crash in the nfa regex engine), but instead of error'ing out it silently fell back to parse the number as a backslash value, instead of the codepoint value and as such this "/[\UFFFFFFFF]" will happily find a "\" or an literal "F".  And this "/[\d127-\UFFFFFFFF]" will error out as "reverse range in character class).

Interestingly, the max Unicode codepoint is U+10FFFF which still fits into an ordinary integer value,  which means, that we don't even need to parse 8 hex characters, but 6 should have been enough. ("6 hex chars ought to be enough for anybody!" anybody?)

However, let's not limit Vim to search for only max 6 hex characters (which would be a backward incompatible change), but instead allow all 8 characters and only if the codepoint is reaches INT_MAX, give a more precise error message (about what the max unicode codepoint value is). This allows to search for "[\U7FFFFFFE]" (will likely return "E486 Pattern not found") and "[/\U7FFFFFF]" now errors "E1517: Value too large, max Unicode codepoint is U+10FFFF".

While this change is straight forward on architectures where long is 8 bytes, this is not so simple on Windows or 32bit achitectures where long is 4 bytes (and therefore the test fails there).  To account for that, let's make use of the vimlong_T number type and make a few corresponding changes in the regex engine code and cast the value to the expected data type. This however may not work correctly on systems that doesn't have the long long datatype (e.g. OpenVMS) and probably the test will fail there.

closes: #16949

zeertzjq commented on src/regexp.c:
```diff
@@ -973,13 +973,14 @@ ungetchr(void)
 /*
  * Get and return the value of the hex string at the current position.
  * Return -1 if there is no valid hex number.
+ * Returns at max INT_MAX
```
```suggestion
 * Returns at maximum INT_MAX.
```

chrisbra replied:
actually, I think the comment should be removed again. I initially added it when trying to limit the return value to MAX_INT, but I removed that again, since it wasn't working as I wanted it to.

chrisbra committed:
regexp: max \U and \%U value is limited by INT_MAX

Problem:  regexp: max \U and \%U value is limited by INT_MAX but gives a
          confusing error message (related: v8.1.0985).
Solution: give a better error message when the value reaches INT_MAX

When searching Vim allows to get up to 8 hex characters using the /\V
and /\%V regex atoms.  However, when using "/\UFFFFFFFF" the code point is
already above what an integer variable can hold, which is 2,147,483,647.

So since patch v8.1.0985, Vim already limited the max codepoint to
INT_MAX (otherwise it caused a crash in the nfa regex engine),
but instead of error'ing out it silently fell back to parse the number
as a backslash value, instead of the codepoint value and as such this
"/[\UFFFFFFFF]" will happily find a "\" or an literal "F".  And this
"/[\d127-\UFFFFFFFF]" will error out as "reverse range in character
class).

Interestingly, the max Unicode codepoint is U+10FFFF which still fits
into an ordinary integer value,  which means, that we don't even need
to parse 8 hex characters, but 6 should have been enough. ("6 hex chars
ought to be enough for anybody!" anybody?)

However, let's not limit Vim to search for only max 6 hex characters
(which would be a backward incompatible change), but instead allow all 8
characters and only if the codepoint is reaches INT_MAX, give a
more precise error message (about what the max unicode codepoint value
is). This allows to search for "[\U7FFFFFFE]" (will likely return "E486
Pattern not found") and "[/\U7FFFFFF]" now errors "E1517: Value too large,
max Unicode codepoint is U+10FFFF".

While this change is straight forward on architectures where long is 8
bytes, this is not so simple on Windows or 32bit achitectures where long
is 4 bytes (and therefore the test fails there).  To account for that,
let's make use of the vimlong_T number type and make a few corresponding
changes in the regex engine code and cast the value to the expected data
type. This however may not work correctly on systems that doesn't have
the long long datatype (e.g. OpenVMS) and probably the test will fail
there.

closes: #16949

Signed-off-by: Christian Brabandt <cb@256bit.org>
